### DISCUSSION TITLE
3. DHCP에 대해 설명해 주세요.

### DISCUSSION THREAD

**Discussion by eun-byeol** [jewelry-box/cs-interview#9] [General]
DHCP에 대해 설명해 주세요.

**Comment by eun-byeol** [comment]
### 🔑 키워드
`IP 주소 및 TCP/IP 설정 자동으로 할당`, `임대`, `중앙 집중식`

---

Dynamic Host Configuration Protocol로, `IP 주소 및 TCP/IP 설정을 자동으로 할당`해주는 UDP 기반 프로토콜입니다. 대부분의 가정용 네트워크에서는 라우터가 DHCP 서버 역할을 합니다.

사용자 PC는 DHCP 서버에서 IP 주소, 기본 게이트웨이(가장 가까운 라우터의 IP 주소), DNS 서버 주소(기본적으로 등록되어 있는 DNS 서버), 서브넷 마스크를 할당받습니다.

`임대` 정책을 통해 자원을 효율적으로 사용합니다. IP 주소는 일정 기간동안 임대되고, 갱신 요청이 없으면 회수합니다. 이를 통해 IP 충돌을 방지하고, IP를 효율적으로 사용할 수 있습니다.

단, `중앙 집중식`으로 관리하다보니 DHCP 서버가 다운되면 새로운 IP 할당이 제대로 이루어지지 않을 수 있습니다.

---
### 더 알아보기

>  DHCP 동작 과정

![image](https://github.com/user-attachments/assets/64247d4e-a172-4348-88a7-4d916316b1c4)

1. `DHCP discover` : 로컬 PC가 브로드캐스트 방식으로 네트워크 상의  DHCP 서버가 어디 있는지 확인한다.
2. `DHCP offer` : DHCP 서버가 로컬 PC에 특정 IP 주소(+ 서브넷 마스크, 게이트웨이, DNS 서버 등 네트워크 설정 정보)를 제안한다.
3. `DHCP Request` : 로컬 PC가 DHCP 서버에게 해당 IP 주소로 할당을 요청한다.
4. `DHCP Ack` : DHCP 서버가 로컬 PC에 해당 IP 주소를 할당한다.

**Comment by cutehumanS2** [comment]
### 추가 질문
**[DHCP는 어느 계층에서 동작하는 프로토콜인가요?]**

DHCP는 애플리케이션 계층에서 동작합니다. 

클라이언트와 서버 간에 IP 주소를 동적으로 할당하기 위해 UDP 프로토콜을 사용해 통신하며, 67번 포트(DHCP 서버)와 68번 포트(DHCP 클라이언트)를 사용합니다.

*+) UDP를 사용하는 이유는 IP 주소를 아직 할당받지 못한 클라이언트는 신뢰성보다는 빠르게 서버와 통신해야 하기 때문이다.*
<br><br>

**[DHCP의 장점과 단점은 무엇인가요?]**

- **장점**
    - 수작업없이 IP 주소를 자동으로 할당받을 수 있어 관리 부담이 줄어듭니다.
    - 중복된 IP 주소 할당을 방지하여, IP 충돌을 최소화합니다.
    - 사용하지 않는 IP 주소를 회수하고 재할당할 수 있어, IP 자원을 효율적으로 관리할 수 있습니다.
- **단점**
    - DHCP 서버가 장애를 일으키면 IP 주소를 할당받을 수 없습니다.
    - IP 주소가 동적으로 변경되기 때문에, 특정 장치가 고정된 IP 주소를 사용해야 하는 경우에는 적합하지 않습니다.
<br><br>

**[DHCP Discover는 왜 브로드캐스트로 보내나요?]**

클라이언트가 IP 주소를 갖고 있지 않기 때문에, 특정 서버를 지정하지 못하고 네트워크 전체에 브로드캐스트로 요청합니다.
<br><br>

**[DHCP에서 IP 주소를 반납하지 않으면 어떻게 되나요?]**

IP 주소가 반납되지 않은 채 장시간 사용되지 않으면, DHCP 서버에서 자동으로 회수되어 다른 장치에 재할당됩니다.
